File: highspeed_report/highspeed_accounts/report/posawesome_opening_closing_report/posawesome_opening_closing_report.js
frappe.query_reports["POSAwesome Opening Closing Report"] = {
    filters: [
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "Link",
            options: "Company",
            default: frappe.defaults.get_user_default("company"),
            reqd: 1
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: frappe.datetime.month_start(),
            reqd: 1
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: frappe.datetime.month_end(),
            reqd: 1
        },
        {
            fieldname: "pos_profile",
            label: __("POS Profile"),
            fieldtype: "Link",
            options: "POS Profile",
            get_query: function() {
                return {
                    filters: {
                        company: frappe.query_report.get_filter_value("company")
                    }
                };
            }
        },
        {
            fieldname: "user",
            label: __("Cashier"),
            fieldtype: "Link",
            options: "User"
        },
        {
            fieldname: "show_payment_details",
            label: __("Show Payment Details"),
            fieldtype: "Check"
        }
    ],

    formatter: function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (data && data.is_total_row) {
            value = `<b>${value}</b>`;
        }

        if (column.fieldname === "difference") {
            const diff = parseFloat(data.difference) || 0;
            if (diff < 0) {
                value = `<span style="color: #e74c3c; font-weight: bold;">${value}</span>`;
            } else if (diff > 0) {
                value = `<span style="color: #27ae60; font-weight: bold;">${value}</span>`;
            }
        }

        if (column.fieldname === "difference_percentage") {
            const percent = parseFloat(data.difference_percentage) || 0;
            if (percent < -5) {
                value = `<span style="color: #e74c3c; font-weight: bold;">${value}</span>`;
            } else if (percent > 5) {
                value = `<span style="color: #e67e22; font-weight: bold;">${value}</span>`;
            }
        }

        if (column.fieldname === "status") {
            if (data.status === "Submitted" || data.status === __("Submitted")) {
                value = `<span class="indicator-pill green">${value}</span>`;
            } else {
                value = `<span class="indicator-pill orange">${value}</span>`;
            }
        }

        return value;
    },

    onload: function(report) {
        report.page.add_inner_button(__("Print"), function() {
            print_custom_report(report);
        }, null, "btn-primary");
    }
};

function print_custom_report(report) {
    const data = frappe.query_report.data || [];
    const columns = frappe.query_report.columns || [];
    const filters = report.get_values();
    const company = filters.company || frappe.defaults.get_user_default("company");
    const from_date = filters.from_date;
    const to_date = filters.to_date;
    const show_details = filters.show_payment_details;

    let html = `
        <!DOCTYPE html>
        <html>
        <head>
            <meta charset="UTF-8">
            <title>${__("POS Opening Closing Report")}</title>
            <style>
                @page {
                    size: A4 landscape;
                    margin: 10mm;
                }
                
                body {
                    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Arial, sans-serif;
                    margin: 0;
                    padding: 20px;
                    color: #2c3e50;
                    background: #fff;
                }
                
                .header {
                    text-align: center;
                    margin-bottom: 30px;
                    border-bottom: 3px solid #3498db;
                    padding-bottom: 20px;
                }
                
                .header h1 {
                    margin: 0;
                    color: #2c3e50;
                    font-size: 28px;
                    font-weight: 600;
                }
                
                .header h2 {
                    margin: 10px 0 0 0;
                    color: #3498db;
                    font-size: 20px;
                    font-weight: 400;
                }
                
                .info-section {
                    display: flex;
                    justify-content: space-between;
                    margin-bottom: 20px;
                    padding: 15px;
                    background: #f8f9fa;
                    border-radius: 8px;
                }
                
                .info-item {
                    display: flex;
                    align-items: center;
                }
                
                .info-label {
                    font-weight: 600;
                    color: #7f8c8d;
                    margin-right: 10px;
                }
                
                .info-value {
                    color: #2c3e50;
                    font-weight: 500;
                }
                
                table {
                    width: 100%;
                    border-collapse: collapse;
                    margin-top: 20px;
                    box-shadow: 0 2px 4px rgba(0,0,0,0.1);
                    border-radius: 8px;
                    overflow: hidden;
                }
                
                thead {
                    background: linear-gradient(135deg, #3498db, #2980b9);
                    color: white;
                }
                
                th {
                    padding: 12px 8px;
                    text-align: center;
                    font-weight: 600;
                    font-size: 12px;
                    letter-spacing: 0.5px;
                    border-right: 1px solid rgba(255,255,255,0.2);
                }
                
                th:last-child {
                    border-right: none;
                }
                
                td {
                    padding: 10px 8px;
                    text-align: center;
                    font-size: 11px;
                    border-bottom: 1px solid #ecf0f1;
                }
                
                tbody tr:nth-child(even) {
                    background-color: #f8f9fa;
                }
                
                tbody tr:hover {
                    background-color: #e8f4f8;
                    transition: background-color 0.2s;
                }
                
                .total-row {
                    background: linear-gradient(135deg, #34495e, #2c3e50) !important;
                    color: white;
                    font-weight: 700;
                    font-size: 13px;
                }
                
                .total-row td {
                    border-bottom: none;
                    padding: 15px 8px;
                }
                
                .negative {
                    color: #e74c3c;
                    font-weight: 600;
                }
                
                .positive {
                    color: #27ae60;
                    font-weight: 600;
                }
                
                .status-submitted {
                    background: #27ae60;
                    color: white;
                    padding: 3px 10px;
                    border-radius: 12px;
                    font-size: 10px;
                    font-weight: 500;
                    display: inline-block;
                }
                
                .status-draft {
                    background: #f39c12;
                    color: white;
                    padding: 3px 10px;
                    border-radius: 12px;
                    font-size: 10px;
                    font-weight: 500;
                    display: inline-block;
                }
                
                .summary {
                    margin-top: 30px;
                    display: grid;
                    grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
                    gap: 15px;
                }
                
                .summary-card {
                    background: #fff;
                    border: 2px solid #ecf0f1;
                    border-radius: 8px;
                    padding: 20px;
                    text-align: center;
                    transition: all 0.3s;
                }
                
                .summary-card:hover {
                    border-color: #3498db;
                    box-shadow: 0 4px 8px rgba(52, 152, 219, 0.1);
                }
                
                .summary-label {
                    color: #7f8c8d;
                    font-size: 12px;
                    font-weight: 500;
                    margin-bottom: 8px;
                }
                
                .summary-value {
                    color: #2c3e50;
                    font-size: 20px;
                    font-weight: 700;
                }
                
                .footer {
                    margin-top: 40px;
                    text-align: center;
                    color: #7f8c8d;
                    font-size: 11px;
                    border-top: 1px solid #ecf0f1;
                    padding-top: 20px;
                }
                
                @media print {
                    body {
                        padding: 0;
                    }
                    
                    .no-print {
                        display: none;
                    }
                    
                    table {
                        box-shadow: none;
                    }
                }
            </style>
        </head>
        <body>
            <div class="header">
                <h1>${company}</h1>
                <h2>${__("POS Opening Closing Report")}</h2>
            </div>
            
            <div class="info-section">
                <div class="info-item">
                    <span class="info-label">${__("Period")}:</span>
                    <span class="info-value">${frappe.datetime.str_to_user(from_date)} - ${frappe.datetime.str_to_user(to_date)}</span>
                </div>
                <div class="info-item">
                    <span class="info-label">${__("Print Date")}:</span>
                    <span class="info-value">${frappe.datetime.now_datetime()}</span>
                </div>
                <div class="info-item">
                    <span class="info-label">${__("Printed By")}:</span>
                    <span class="info-value">${frappe.session.user_fullname}</span>
                </div>
            </div>
            
            <table>
                <thead>
                    <tr>`;

    columns.forEach(col => {
        if (!show_details && col.fieldname === "mode_of_payment") return;
        html += `<th>${col.label}</th>`;
    });

    html += `</tr></thead><tbody>`;

    data.forEach(row => {
        const is_total = row.is_total_row;
        html += `<tr${is_total ? ' class="total-row"' : ''}>`;
        
        columns.forEach(col => {
            if (!show_details && col.fieldname === "mode_of_payment") return;
            
            let val = row[col.fieldname];
            let formatted_val = val || "";
            
            if (col.fieldtype === "Currency" && val !== null && val !== undefined) {
                formatted_val = format_currency(val, row.currency);
            } else if (col.fieldtype === "Percent" && val !== null && val !== undefined) {
                formatted_val = val.toFixed(2) + "%";
            } else if (col.fieldtype === "Datetime" && val) {
                formatted_val = frappe.datetime.str_to_user(val);
            }
            
            if (col.fieldname === "difference" && val) {
                const diff = parseFloat(val);
                if (diff < 0) {
                    formatted_val = `<span class="negative">${formatted_val}</span>`;
                } else if (diff > 0) {
                    formatted_val = `<span class="positive">${formatted_val}</span>`;
                }
            }
            
            if (col.fieldname === "status" && val) {
                if (val === "Submitted" || val === __("Submitted")) {
                    formatted_val = `<span class="status-submitted">${val}</span>`;
                } else {
                    formatted_val = `<span class="status-draft">${val}</span>`;
                }
            }
            
            html += `<td>${formatted_val}</td>`;
        });
        
        html += `</tr>`;
    });

    html += `</tbody></table>`;

    // Always show summary totals
    let total_row = data.find(r => r.is_total_row);
    
    // If no total row exists, calculate it
    if (!total_row && data.length > 0) {
        total_row = {
            opening_amount: 0,
            sales_amount: 0,
            expected_amount: 0,
            closing_amount: 0,
            difference: 0,
            currency: data[0].currency || "SAR"
        };
        
        data.forEach(row => {
            if (!row.is_total_row) {
                total_row.opening_amount += parseFloat(row.opening_amount) || 0;
                total_row.sales_amount += parseFloat(row.sales_amount) || 0;
                total_row.expected_amount += parseFloat(row.expected_amount) || 0;
                total_row.closing_amount += parseFloat(row.closing_amount) || 0;
                total_row.difference += parseFloat(row.difference) || 0;
            }
        });
    }
    
    if (total_row) {
        html += `
            <div class="summary">
                <div class="summary-card">
                    <div class="summary-label">${__("Total Opening")}</div>
                    <div class="summary-value">${format_currency(total_row.opening_amount, total_row.currency)}</div>
                </div>
                <div class="summary-card">
                    <div class="summary-label">${__("Total Sales")}</div>
                    <div class="summary-value">${format_currency(total_row.sales_amount, total_row.currency)}</div>
                </div>
                <div class="summary-card">
                    <div class="summary-label">${__("Total Expected")}</div>
                    <div class="summary-value">${format_currency(total_row.expected_amount, total_row.currency)}</div>
                </div>
                <div class="summary-card">
                    <div class="summary-label">${__("Total Closing")}</div>
                    <div class="summary-value">${format_currency(total_row.closing_amount, total_row.currency)}</div>
                </div>
                <div class="summary-card">
                    <div class="summary-label">${__("Total Difference")}</div>
                    <div class="summary-value ${total_row.difference < 0 ? 'negative' : 'positive'}">${format_currency(total_row.difference, total_row.currency)}</div>
                </div>
            </div>`;
    }

    html += `
            <div class="footer">
                <p>${company} - ${__("HIGH SPEED IT REPORT")}</p>
            </div>
        </body>
        </html>`;

    const printWindow = window.open("", "_blank");
    printWindow.document.write(html);
    printWindow.document.close();
    printWindow.focus();
    
    setTimeout(() => {
        printWindow.print();
    }, 250);
}

function format_currency(value, currency) {
    if (!value && value !== 0) return "";
    
    const number = parseFloat(value);
    if (isNaN(number)) return value;
    
    // Format with commas and 2 decimal places
    const formatted = number.toLocaleString(undefined, {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2
    });
    
    return formatted;
}

File: highspeed_report/highspeed_accounts/report/posawesome_opening_closing_report/posawesome_opening_closing_report.py
import frappe
from frappe.utils import flt, get_datetime, fmt_money, getdate
from frappe import _


def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    
    # Get the total row from data for report summary
    total_row = None
    for row in data:
        if row.get("is_total_row"):
            total_row = row
            break
    
    # Create report summary if we have totals
    report_summary = None
    if total_row:
        report_summary = [
            {
                "value": total_row.get("opening_amount", 0),
                "label": _("Total Opening"),
                "indicator": "blue",
                "datatype": "Currency"
            },
            {
                "value": total_row.get("sales_amount", 0),
                "label": _("Total Sales"),
                "indicator": "green",
                "datatype": "Currency"
            },
            {
                "value": total_row.get("expected_amount", 0),
                "label": _("Total Expected"),
                "indicator": "orange",
                "datatype": "Currency"
            },
            {
                "value": total_row.get("closing_amount", 0),
                "label": _("Total Closing"),
                "indicator": "purple",
                "datatype": "Currency"
            },
            {
                "value": abs(total_row.get("difference", 0)),
                "label": _("Total Difference"),
                "indicator": "red" if total_row.get("difference", 0) < 0 else "green",
                "datatype": "Currency"
            }
        ]
    
    return columns, data, None, None, report_summary


def get_doctype_names():
    doctype_opening = "POS Opening Shift"
    doctype_closing = "POS Closing Shift"
    doctype_opening_detail = "POS Opening Shift Detail"
    doctype_closing_detail = "POS Closing Shift Detail"
    fk_opening_shift = "pos_opening_shift"
    
    # Fallback to highspeed_pos doctypes if posawesome is not installed
    if not frappe.db.exists("DocType", doctype_opening) and frappe.db.exists("DocType", "HSPOS Opening Shift"):
        doctype_opening = "HSPOS Opening Shift"
        doctype_closing = "HSPOS Closing Shift"
        doctype_opening_detail = "HSPOS Opening Shift Detail"
        doctype_closing_detail = "HSPOS Closing Shift Detail"
        fk_opening_shift = "hspos_opening_shift"
        
    return doctype_opening, doctype_closing, doctype_opening_detail, doctype_closing_detail, fk_opening_shift


def get_columns(filters):
    doctype_opening, doctype_closing, _, _, _ = get_doctype_names()
    
    columns = [
        {
            "label": _("Opening Shift"),
            "fieldname": "opening_shift",
            "fieldtype": "Link",
            "options": doctype_opening,
            "width": 180
        },
        {
            "label": _("Closing Shift"),
            "fieldname": "closing_shift",
            "fieldtype": "Link",
            "options": doctype_closing,
            "width": 180
        },
        {
            "label": _("Start Date"),
            "fieldname": "start_date",
            "fieldtype": "Datetime",
            "width": 150
        },
        {
            "label": _("End Date"),
            "fieldname": "end_date",
            "fieldtype": "Datetime",
            "width": 150
        },
        {
            "label": _("POS Profile"),
            "fieldname": "pos_profile",
            "fieldtype": "Link",
            "options": "POS Profile",
            "width": 150
        },
        {
            "label": _("Cashier"),
            "fieldname": "cashier",
            "fieldtype": "Link",
            "options": "User",
            "width": 120
        },
        {
            "label": _("Status"),
            "fieldname": "status",
            "fieldtype": "Data",
            "width": 100
        }
    ]
    
    if filters.get("show_payment_details"):
        columns.append({
            "label": _("Mode of Payment"),
            "fieldname": "mode_of_payment",
            "fieldtype": "Link",
            "options": "Mode of Payment",
            "width": 140
        })
    
    columns.extend([
        {
            "label": _("Opening Amount"),
            "fieldname": "opening_amount",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Sales Amount"),
            "fieldname": "sales_amount",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Expected Amount"),
            "fieldname": "expected_amount",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Closing Amount"),
            "fieldname": "closing_amount",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Difference"),
            "fieldname": "difference",
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "label": _("Difference %"),
            "fieldname": "difference_percentage",
            "fieldtype": "Percent",
            "width": 100
        }
    ])
    
    return columns


def get_data(filters):
    rows = []
    filters = filters or {}
    from_date = filters.pop("from_date", None)
    to_date = filters.pop("to_date", None)
    show_payment_details = filters.pop("show_payment_details", 0)
    pos_profile_filter = filters.pop("pos_profile", None)
    user_filter = filters.pop("user", None)
    company = filters.get("company", frappe.defaults.get_user_default("company"))

    doctype_opening, doctype_closing, doctype_opening_detail, doctype_closing_detail, fk_opening_shift = get_doctype_names()

    if from_date:
        from_date = get_datetime(from_date)
    if to_date:
        to_date = get_datetime(to_date)

    closing_filters = {
        "docstatus": ["!=", 2]
    }
    
    if pos_profile_filter:
        closing_filters["pos_profile"] = pos_profile_filter
    if user_filter:
        closing_filters["user"] = user_filter
    if company:
        closing_filters["company"] = company

    closing_shifts = frappe.get_all(
        doctype_closing,
        filters=closing_filters,
        fields=[
            "name", fk_opening_shift, "period_start_date", 
            "period_end_date", "user", "pos_profile", "docstatus",
            "posting_date", "company"
        ]
    )

    if from_date or to_date:
        closing_shifts = [
            c for c in closing_shifts
            if (not from_date or get_datetime(c.period_end_date) >= from_date)
            and (not to_date or get_datetime(c.period_end_date) <= to_date)
        ]

    currency = frappe.get_cached_value("Company", company, "default_currency") if company else "SAR"
    
    # Track processed data to avoid duplicates
    processed_shifts = set()

    for closing in closing_shifts:
        opening_name = closing.get(fk_opening_shift)
        if not opening_name:
            continue
        
        # Skip if already processed (in case of duplicates)
        shift_key = f"{opening_name}-{closing.name}"
        if shift_key in processed_shifts:
            continue
        processed_shifts.add(shift_key)
        
        opening_data = frappe.db.get_value(
            doctype_opening, 
            opening_name, 
            ["period_start_date", "user"],
            as_dict=True
        )
        
        if not opening_data:
            continue

        if show_payment_details:
            opening_details = frappe.get_all(
                doctype_opening_detail,
                filters={"parent": opening_name},
                fields=["mode_of_payment", "amount"]
            )
            opening_map = {d.mode_of_payment: flt(d.amount) for d in opening_details}

            closing_details = frappe.get_all(
                doctype_closing_detail,
                filters={"parent": closing.name},
                fields=[
                    "mode_of_payment", "expected_amount", 
                    "closing_amount", "difference"
                ]
            )
            
            sales_by_mode = get_sales_by_mode(opening_name, closing.name, doctype_opening_detail, doctype_closing_detail)
            
            closing_map = {d.mode_of_payment: d for d in closing_details}

            modes = set(opening_map.keys()) | set(closing_map.keys())
            for mop in modes:
                closing_row = closing_map.get(mop, {})
                expected_amt = flt(closing_row.get("expected_amount", 0))
                closing_amt = flt(closing_row.get("closing_amount", 0))
                diff = flt(closing_row.get("difference", 0))
                diff_percent = (diff / expected_amt * 100) if expected_amt else 0
                
                row = {
                    "opening_shift": opening_name,
                    "closing_shift": closing.name,
                    "start_date": opening_data.period_start_date,
                    "end_date": closing.period_end_date,
                    "pos_profile": closing.pos_profile,
                    "cashier": closing.user,
                    "status": _("Submitted") if closing.docstatus == 1 else _("Draft"),
                    "mode_of_payment": mop,
                    "opening_amount": flt(opening_map.get(mop, 0)),
                    "sales_amount": flt(sales_by_mode.get(mop, 0)),
                    "expected_amount": expected_amt,
                    "closing_amount": closing_amt,
                    "difference": diff,
                    "difference_percentage": diff_percent,
                    "currency": currency
                }
                rows.append(row)
        else:
            closing_totals = frappe.db.sql(f"""
                SELECT 
                    SUM(expected_amount) as expected_amount,
                    SUM(closing_amount) as closing_amount,
                    SUM(difference) as difference
                FROM `tab{doctype_closing_detail}`
                WHERE parent = %s
            """, closing.name, as_dict=True)[0]
            
            opening_total = frappe.db.sql(f"""
                SELECT SUM(amount) as amount
                FROM `tab{doctype_opening_detail}`
                WHERE parent = %s
            """, opening_name, as_dict=True)[0]
            
            total_sales = get_total_sales(opening_name, closing.name, doctype_opening_detail, doctype_closing_detail, doctype_closing)
            
            expected_amt = flt(closing_totals.expected_amount)
            closing_amt = flt(closing_totals.closing_amount)
            diff = flt(closing_totals.difference)
            diff_percent = (diff / expected_amt * 100) if expected_amt else 0

            row = {
                "opening_shift": opening_name,
                "closing_shift": closing.name,
                "start_date": opening_data.period_start_date,
                "end_date": closing.period_end_date,
                "pos_profile": closing.pos_profile,
                "cashier": closing.user,
                "status": _("Submitted") if closing.docstatus == 1 else _("Draft"),
                "opening_amount": flt(opening_total.amount),
                "sales_amount": flt(total_sales),
                "expected_amount": expected_amt,
                "closing_amount": closing_amt,
                "difference": diff,
                "difference_percentage": diff_percent,
                "currency": currency
            }
            rows.append(row)

    # Calculate totals properly
    if rows:
        # Always calculate fresh totals
        total_opening = 0
        total_sales = 0
        total_expected = 0
        total_closing = 0
        total_difference = 0
        
        for row in rows:
            if not row.get("is_total_row"):
                total_opening += flt(row.get("opening_amount", 0))
                total_sales += flt(row.get("sales_amount", 0))
                total_expected += flt(row.get("expected_amount", 0))
                total_closing += flt(row.get("closing_amount", 0))
                total_difference += flt(row.get("difference", 0))
        
        total_row = {
            "opening_shift": _("Total"),
            "closing_shift": "",
            "opening_amount": total_opening,
            "sales_amount": total_sales,
            "expected_amount": total_expected,
            "closing_amount": total_closing,
            "difference": total_difference,
            "is_total_row": True,
            "currency": currency
        }
        
        if show_payment_details:
            total_row["mode_of_payment"] = ""
        
        if total_expected > 0:
            total_row["difference_percentage"] = (total_difference / total_expected * 100)
        else:
            total_row["difference_percentage"] = 0
            
        rows.append(total_row)

    return rows


def get_sales_by_mode(opening_shift, closing_shift, doctype_opening_detail="POS Opening Shift Detail", doctype_closing_detail="POS Closing Shift Detail"):
    # Get from POS Closing Shift Payment details
    payment_details = frappe.get_all(
        doctype_closing_detail,
        filters={"parent": closing_shift},
        fields=["mode_of_payment", "expected_amount"]
    )
    
    if payment_details:
        # Get opening amounts
        opening_details = frappe.get_all(
            doctype_opening_detail,
            filters={"parent": opening_shift},
            fields=["mode_of_payment", "amount"]
        )
        opening_map = {d.mode_of_payment: flt(d.amount) for d in opening_details}
        
        sales_by_mode = {}
        for payment in payment_details:
            mode = payment.mode_of_payment
            expected = flt(payment.expected_amount)
            opening = opening_map.get(mode, 0)
            # Sales = Expected - Opening
            sales_by_mode[mode] = expected - opening
        
        return sales_by_mode
    
    return {}


def get_total_sales(opening_shift, closing_shift, doctype_opening_detail="POS Opening Shift Detail", doctype_closing_detail="POS Closing Shift Detail", doctype_closing="POS Closing Shift"):
    # Get sales from POS Closing Shift directly
    closing_data = frappe.db.get_value(
        doctype_closing,
        closing_shift,
        ["grand_total", "net_total"],
        as_dict=True
    )
    
    if closing_data and closing_data.get("grand_total"):
        return flt(closing_data.grand_total)
    
    # If no grand_total in closing shift, calculate from expected amounts
    expected_total = frappe.db.sql(f"""
        SELECT SUM(expected_amount) as total
        FROM `tab{doctype_closing_detail}`
        WHERE parent = %s
    """, closing_shift)
    
    opening_total = frappe.db.sql(f"""
        SELECT SUM(amount) as total
        FROM `tab{doctype_opening_detail}`
        WHERE parent = %s
    """, opening_shift)
    
    expected = flt(expected_total[0][0]) if expected_total and expected_total[0][0] else 0
    opening = flt(opening_total[0][0]) if opening_total and opening_total[0][0] else 0
    
    # Sales = Expected - Opening
    return expected - opening